### PR TITLE
Improvements for PaginatorHelper

### LINKED ISSUE #9503: Could not generate pagination links with passed parameters in route.
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.3
* Platform and Target: Apache 2.4, PHP 7.0.10, MySQL

### What you did

Configured this route:

```
Router::scope('/:lang', [
    'lang' => '[a-z][a-z]'
    ], function (RouteBuilder $routes) {
    $routes->connect('/explore', [
        'controller' => 'Explore',
        'action' => 'index'
    ], [
        '_name' => 'explore',
        'pass' => ['lang']
    ]);
});
```

Tried to generate pagination numbers:

```
echo $this->Paginator->numbers(['url' => ['lang' => 'en']]);
```

Triggered this missing route error:

```
Error: A route matching "array ( 'page' => false, 0 => 'en', 'lang' => 'en', 'action' => 'index', 'plugin' => NULL, 'controller' => 'Explore', '_ext' => NULL, )" could not be found.
```

Notice that both the passed parameter `lang` is defined twice. Once as `0=>'en'` and `'lang'=>'en'`.

### Expected Behavior

It should not be inject named parameters as a numeric key.

### Actual Behavior

The source of the problem was in the PaginatorHelper constructor.

```
        $this->config(
            'options.url',
            array_merge($this->request->params['pass'], ['?' => $query])
        );
```
The above is attempting to persist the current route parameters. Which is a good feature, but I think injecting `params['pass']` is the wrong approach. Since those are numeric key/pair set.

This raises the question of where the bug is located. Should the `params['pass']` contain keys with the name of the parameter instead of numeric keys, or should the helper's constructor inject named parameter manually (some how).

The work around for the above problem was to erase the options URL before generating links.

```
        $this->Paginator->config('options.url',[], false);
```

### PR DESCRIPTION
**Pull request by thinkingmedia**
Adds the following to the helper
- moved core of generateUrl() to buildUrl() and returns a raw URL array of pagination parameters.
- generateUrl calls buildUrl()
- added total() which returns page count.
- added to meta() support for generating first and last meta links
- The object of the PR is to allow developers to generate pagination URLs in their array format for easier modification outside of the helper.

closes #9503

### CODE REVIEW

File: tests/TestCase/View/Helper/PaginatorHelperTest.php [hunk 1]
```diff
@@ -44,9 +54,6 @@ public function setUp()
         Configure::write('Config.language', 'eng');
         $this->View = new View();
         $this->Paginator = new PaginatorHelper($this->View);
-        $this->Paginator->Js = $this->getMockBuilder('Cake\View\Helper\PaginatorHelper')
-            ->setConstructorArgs([$this->View])
-            ->getMock();
```
Comment (markstory) [inline tests/TestCase/View/Helper/PaginatorHelperTest.php]: :clap:

**Comment by markstory** [comment]
Looks good to me.

File: src/View/Helper/PaginatorHelper.php [hunk 2]
```diff
@@ -1076,40 +1106,58 @@ public function last($last = 'last >>', array $options = [])
...
+     * - `next` (default True) True to generate meta for next page
+     * - `first` (default False) True to generate meta for first page
+     * - `last` (default False) True to generate meta for last page
      *
      * @param array $options Array of options
      * @return string|null Meta links
      */
     public function meta(array $options = [])
     {
+        $options = $options + [
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: no need for doubling:

    $options += [

File: src/View/Helper/PaginatorHelper.php [hunk 3]
```diff
@@ -1076,40 +1106,58 @@ public function last($last = 'last >>', array $options = [])
...
         }
 
+        $out = implode($links);
+
         if ($options['block'] === true) {
             $options['block'] = __FUNCTION__;
         }
-        $this->_View->append($options['block'], $out);
+
+        if (!empty($options['block'])) {
```
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: `if ($options['block'])` as it must be defined

File: src/View/Helper/PaginatorHelper.php (hunk 2 above)
Comment (dereuromark) [inline src/View/Helper/PaginatorHelper.php]: I am 100% sure it does. It is the same operation.

Comment (thinkingmedia) [inline src/View/Helper/PaginatorHelper.php]: okay, I'll change it.

**Comment by dereuromark** [comment]
Looks good to me, too.

**Pull request merged**